vortex-file/src/read/layouts/inline_dtype.rs

```diff
@@ -91,45 +80,40 @@ impl InlineDTypeLayout {
         }
     }
 
-    fn dtype(&self) -> VortexResult<DTypeReadResult> {
-        if let Some(dt_bytes) = self.message_cache.get(&[INLINE_DTYPE_BUFFER_IDX]) {
-            let msg = root::<message::Message>(&dt_bytes[MESSAGE_PREFIX_LENGTH..])?
-                .header_as_schema()
-                .ok_or_else(|| vortex_err!("Expected schema message"))?;
-
-            Ok(DTypeReadResult::DType(DType::try_from(
-                msg.dtype()
-                    .ok_or_else(|| vortex_err!(InvalidSerde: "Schema missing DType"))?,
-            )?))
-        } else {
-            let buffers = self.flatbuffer().buffers().unwrap_or_default();
-            if buffers.is_empty() {
-                vortex_bail!("Missing buffers for inline dtype layout")
-            }
-            let dtype_buf = buffers.get(0);
-            Ok(DTypeReadResult::ReadMore(vec![MessageLocator(
-                self.message_cache.absolute_id(&[INLINE_DTYPE_BUFFER_IDX]),
-                ByteRange::new(dtype_buf.begin(), dtype_buf.end()),
-            )]))
+    fn dtype_message(&self) -> VortexResult<MessageLocator> {
+        let buffers = self.flatbuffer().buffers().unwrap_or_default();
+        if buffers.is_empty() {
+            vortex_bail!("Missing buffers for inline dtype layout")
         }
+        let dtype_buf = buffers.get(0);
+        Ok(MessageLocator(
+            self.message_cache.absolute_id(&[INLINE_DTYPE_BUFFER_IDX]),
+            ByteRange::new(dtype_buf.begin(), dtype_buf.end()),
+        ))
     }
 
-    fn child_reader(&self) -> VortexResult<ChildReaderResult> {
-        match self.dtype()? {
-            DTypeReadResult::ReadMore(m) => Ok(ChildReaderResult::ReadMore(m)),
-            DTypeReadResult::DType(d) => {
-                let child_layout = self.layout_builder.read_layout(
-                    self.fb_bytes.clone(),
-                    self.child_layout()?._tab.loc(),
-                    self.scan.clone(),
-                    self.message_cache.relative(
-                        INLINE_DTYPE_CHILD_IDX,
-                        Arc::new(LazilyDeserializedDType::from_dtype(d)),
-                    ),
-                )?;
-                Ok(ChildReaderResult::Reader(child_layout))
-            }
-        }
+    fn dtype(&self) -> Arc<LazilyDeserializedDType> {
+        Arc::new(unsafe {
+            LazilyDeserializedDType::from_lazy_schema_bytes({
+                let cache = self.message_cache.clone();
+                move || {
+                    cache
+                        .get(&[INLINE_DTYPE_BUFFER_IDX])
+                        .vortex_expect("Must have populated message cache with dtype message")
+                        .slice(MESSAGE_PREFIX_LENGTH..)
+                }
+            })
```

Comment:
It’s an artefact of how io here is done. It would be nice to be able to eagerly construct layout readers but this one needs to do io to construct its children. I think since we still do a once cell we can delegate that logic there.

Reply:
If we didn’t need names in columnar layout to evaluate expression we could avoid the oncecell but I think it would be quite hard to do nicely